File: phoenix/lib/cards_project_web/controllers/cards/card_set_controller.ex
defmodule CardsProjectWeb.Cards.CardSetController do
  use Phoenix.Controller, formats: [:json]
  import Plug.Conn

  alias CardsProject.Cards
  alias CardsProject.Cards.CardSet

  def index(conn, _params) do
    card_sets = Cards.list_card_sets()
    json(conn, Enum.map(card_sets, &serialize_card_set/1))
  end

  def show(conn, %{"id" => id}) do
    card_set = Cards.get_card_set!(id)
    json(conn, serialize_card_set(card_set))
  end

  def create(conn, params) do
    case Cards.create_card_set(params) do
      {:ok, card_set} ->
        conn
        |> put_status(:created)
        |> json(serialize_card_set(card_set))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  def update(conn, %{"id" => id} = params) do
    card_set = Cards.get_card_set!(id)
    case Cards.update_card_set(card_set, params) do
      {:ok, card_set} ->
        json(conn, serialize_card_set(card_set))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  def delete(conn, %{"id" => id}) do
    card_set = Cards.get_card_set!(id)
    Cards.delete_card_set(card_set)
    send_resp(conn, :no_content, "")
  end

  # GET /api/card-sets/{id}/standard-legal
  def is_legal_in_standard(conn, %{"id" => id}) do
    result = Cards.card_set_is_legal_in_standard_behavior(id)
    json(conn, %{result: result})
  end

  # GET /api/card-sets/{id}/legal
  def is_legal_in_format(conn, %{"id" => id} = params) do
    format = Map.get(params, "format")
    result = Cards.card_set_is_legal_in_format_behavior(id, format)
    json(conn, %{result: result})
  end

  # GET /api/card-sets/{id}/rarity-count
  def card_count_by_rarity(conn, %{"id" => id} = params) do
    rarity = Map.get(params, "rarity")
    result = Cards.card_set_card_count_by_rarity_behavior(id, rarity)
    json(conn, %{result: result})
  end

  # POST /api/card-sets/{id}/rotate
  def rotate_out(conn, %{"id" => id}) do
    Cards.card_set_rotate_out_behavior(id)
    send_resp(conn, :no_content, "")
  end

  defp serialize_card_set(%CardSet{} = record) do
    Map.take(record, [:id, :name, :code, :release_date, :rotation_date, :set_type, :total_cards, :is_rotated, :description, :logo_url])
  end

  defp format_errors(changeset) do
    Ecto.Changeset.traverse_errors(changeset, fn {msg, opts} ->
      Regex.replace(~r"%{(\w+)}", msg, fn _, key ->
        opts |> Keyword.get(String.to_existing_atom(key), key) |> to_string()
      end)
    end)
  end
end

File: symfony/src/Controller/Api/Cards/CardSetController.php
<?php

namespace App\Controller\Api\Cards;

use App\Entity\Cards\CardSet;
use App\Repository\Cards\CardSetRepository;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Attribute\Route;
use Symfony\Component\Validator\Validator\ValidatorInterface;


#[Route('/api/card_sets', name: 'cardSet_')]
class CardSetController extends AbstractController
{
    public function __construct(
        private CardSetRepository $repository,
        private ValidatorInterface $validator,
    ) {}

    #[Route('', name: 'list', methods: ['GET'])]
    public function list(): JsonResponse
    {
        $items = $this->repository->findAll();
        return $this->json($items, context: ['groups' => ['cardSet:read']]);
    }

    #[Route('', name: 'create', methods: ['POST'])]
    public function create(Request $request): JsonResponse
    {
        $data = json_decode($request->getContent(), true) ?? [];
        $cardSet = new CardSet();
        if (isset($data['name'])) $cardSet->setName($data['name']);
        if (isset($data['code'])) $cardSet->setCode($data['code']);
        if (isset($data['releaseDate'])) $cardSet->setReleaseDate(new \DateTime($data['releaseDate']));
        if (isset($data['rotationDate'])) $cardSet->setRotationDate(new \DateTime($data['rotationDate']));
        if (isset($data['setType'])) $cardSet->setSetType($data['setType']);
        if (isset($data['totalCards'])) $cardSet->setTotalCards($data['totalCards']);
        if (isset($data['isRotated'])) $cardSet->setIsRotated($data['isRotated']);
        if (isset($data['description'])) $cardSet->setDescription($data['description']);
        if (isset($data['logoUrl'])) $cardSet->setLogoUrl($data['logoUrl']);


        $errors = $this->validator->validate($cardSet);
        if (count($errors) > 0) {
            return $this->json(['errors' => (string) $errors], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        try {
            $cardSet->validateImplies();
        } catch (\DomainException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        $this->repository->save($cardSet, flush: true);
        return $this->json($cardSet, Response::HTTP_CREATED, context: ['groups' => ['cardSet:read']]);
    }

    #[Route('/{id}', name: 'show', methods: ['GET'])]
    public function show(CardSet $cardSet): JsonResponse
    {
        return $this->json($cardSet, context: ['groups' => ['cardSet:read']]);
    }

    #[Route('/{id}', name: 'update', methods: ['PUT', 'PATCH'])]
    public function update(Request $request, CardSet $cardSet): JsonResponse
    {
        $data = json_decode($request->getContent(), true) ?? [];
        if (isset($data['name'])) $cardSet->setName($data['name']);
        if (isset($data['code'])) $cardSet->setCode($data['code']);
        if (isset($data['releaseDate'])) $cardSet->setReleaseDate(new \DateTime($data['releaseDate']));
        if (isset($data['rotationDate'])) $cardSet->setRotationDate(new \DateTime($data['rotationDate']));
        if (isset($data['setType'])) $cardSet->setSetType($data['setType']);
        if (isset($data['totalCards'])) $cardSet->setTotalCards($data['totalCards']);
        if (isset($data['isRotated'])) $cardSet->setIsRotated($data['isRotated']);
        if (isset($data['description'])) $cardSet->setDescription($data['description']);
        if (isset($data['logoUrl'])) $cardSet->setLogoUrl($data['logoUrl']);


        $errors = $this->validator->validate($cardSet);
        if (count($errors) > 0) {
            return $this->json(['errors' => (string) $errors], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        try {
            $cardSet->validateImplies();
        } catch (\DomainException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        $this->repository->save($cardSet, flush: true);
        return $this->json($cardSet, context: ['groups' => ['cardSet:read']]);
    }

    #[Route('/{id}', name: 'delete', methods: ['DELETE'])]
    public function delete(CardSet $cardSet): JsonResponse
    {
        $this->repository->remove($cardSet, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }

    #[Route('/{id}/standard-legal', name: 'isLegalInStandard', methods: ['GET'])]
    public function isLegalInStandard(CardSet $cardSet): JsonResponse
    {
        $result = $cardSet->isLegalInStandard();
        $this->repository->save($cardSet, flush: true);
        return $this->json($result);
    }

    #[Route('/{id}/legal', name: 'isLegalInFormat', methods: ['GET'])]
    public function isLegalInFormat(CardSet $cardSet): JsonResponse
    {
        $result = $cardSet->isLegalInFormat($format);
        $this->repository->save($cardSet, flush: true);
        return $this->json($result);
    }

    #[Route('/{id}/rarity-count', name: 'cardCountByRarity', methods: ['GET'])]
    public function cardCountByRarity(CardSet $cardSet): JsonResponse
    {
        $result = $cardSet->cardCountByRarity($rarity);
        $this->repository->save($cardSet, flush: true);
        return $this->json($result);
    }

    #[Route('/{id}/rotate', name: 'rotateOut', methods: ['POST'])]
    public function rotateOut(CardSet $cardSet): JsonResponse
    {
        $cardSet->rotateOut();
        $this->repository->save($cardSet, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }
}
